### ISSUE TITLE
NSManagedObjectContext Multithreading_Violation_AllThatIsLeftToUsIsHonor__

### ISSUE THREAD

**Issue by andreyz** [helpscout/beacon-ios-sdk#105]
* Beacon install method: [ ] Carthage, [ ] CocoaPods, [x] Manual

**Actual outcome**
CoreData multithreading violation triggered on minmizing the chat.
```
#0	0x00000001b781319c in +[NSManagedObjectContext __Multithreading_Violation_AllThatIsLeftToUsIsHonor__] ()
#1	0x00000001b7813080 in _PFAssertSafeMultiThreadedAccess_impl ()
#2	0x00000001b76f0da4 in -[NSManagedObjectContext executeFetchRequest:error:] ()
#3	0x00000001f6cd8504 in NSManagedObjectContext.fetch<A>(_:) ()
#4	0x000000010605b2c4 in specialized static Chat.object(from:in:) at /Users/travis/build/helpscout/beacon-ios-sdk-private/Beacon/Classes/Chat/Value Types/Chat.swift:19
#5	0x00000001060c15d4 in static Chat.object(from:in:) [inlined] ()
#6	0x00000001060c15c8 in PersistentContainer.persist(_:) [inlined] at /Users/travis/build/helpscout/beacon-ios-sdk-private/Beacon/Classes/Chat/PersistentContainer.swift:18
#7	0x00000001060c15a4 in static ChatReducer.endChatReducer(state:action:) at /Users/travis/build/helpscout/beacon-ios-sdk-private/Beacon/Classes/Chat/ChatReducer.swift:437
#8	0x00000001060d14d0 in thunk for @escaping @callee_guaranteed (@inout ChatState, @in_guaranteed ChatViewAction) -> (@owned [Effect<ChatViewAction>]) [inlined] ()
#9	0x00000001060d14cc in partial apply for thunk for @escaping @callee_guaranteed (@inout ChatState, @in_guaranteed ChatViewAction) -> (@owned [Effect<ChatViewAction>]) ()
#10	0x0000000106066650 in specialized thunk for @escaping @callee_guaranteed (@inout A, @in_guaranteed B) -> (@out [Effect<B>]) [inlined] ()
#11	0x0000000106066640 in specialized closure #1 in closure #1 in combine<A, B>(_:) [inlined] at /Users/travis/build/helpscout/beacon-ios-sdk-private/Beacon/Classes/Chat/ChatArchitecture.swift:76
#12	0x0000000106066640 in specialized thunk for @callee_guaranteed (@guaranteed @escaping @callee_guaranteed (@inout A, @in_guaranteed B) -> (@owned [Effect<B>])) -> (@owned [Effect<B>], @error @owned Error) [inlined] ()
#13	0x0000000106066640 in specialized Sequence.flatMap<A>(_:) [inlined] ()
#14	0x0000000106066640 in specialized closure #1 in combine<A, B>(_:) at /Users/travis/build/helpscout/beacon-ios-sdk-private/Beacon/Classes/Chat/ChatArchitecture.swift:76
#15	0x0000000106065d94 in specialized Store.reduce(_:) [inlined] at /Users/travis/build/helpscout/beacon-ios-sdk-private/Beacon/Classes/Chat/ChatArchitecture.swift:24
#16	0x0000000106065d88 in specialized closure #1 in closure #1 in Store.reduce(_:) at /Users/travis/build/helpscout/beacon-ios-sdk-private/Beacon/Classes/Chat/ChatArchitecture.swift:27
#17	0x00000001060ff68c in specialized closure #1 in closure #1 in Effect.map<A>(_:) at /Users/travis/build/helpscout/beacon-ios-sdk-private/Beacon/Classes/Chat/Effect.swift:11
#18	0x00000001060434b4 in specialized closure #1 in closure #1 in APIClient.load<A>(_:) [inlined] at /Users/travis/build/helpscout/beacon-ios-sdk-private/Beacon/Classes/Root/Swift Networking/APIClient.swift:94
#19	0x00000001060434b0 in partial apply for specialized closure #1 in closure #1 in APIClient.load<A>(_:) ()
#20	0x0000000106039ed0 in specialized APIClient.handle<A>(response:for:with:) at /Users/travis/build/helpscout/beacon-ios-sdk-private/Beacon/Classes/Root/Swift Networking/APIClient.swift:244
#21	0x000000010604299c in partial apply for specialized closure #2 in APIClient.load<A>(_:completion:) ()
#22	0x00000001060312dc in thunk for @escaping @callee_guaranteed (@guaranteed Data?, @guaranteed NSURLResponse?, @guaranteed Error?) -> () ()
#23	0x00000001b1fc4540 in ___lldb_unnamed_symbol531$$CFNetwork ()
#24	0x00000001b1fd5010 in ___lldb_unnamed_symbol921$$CFNetwork ()
#25	0x00000001078ebdb8 in _dispatch_call_block_and_release ()
#26	0x00000001078ed5fc in _dispatch_client_callout ()
#27	0x00000001078f4680 in _dispatch_lane_serial_drain ()
#28	0x00000001078f533c in _dispatch_lane_invoke ()
#29	0x0000000107900b34 in _dispatch_workloop_worker_thread ()
#30	0x00000001f6fb05a4 in _pthread_wqthread ()
```

**Expected outcome**
Beacon shouldn't trigger CoreData multithreading errors. Please add `-com.apple.CoreData.ConcurrencyDebug 1` to your Launch Arguments and verify.

**Comment by andreyz** [comment]
P.S. Are you using [TCA](https://github.com/pointfreeco/swift-composable-architecture) for implementing the chat feature?

**Comment by ScottPetit** [comment]
> P.S. Are you using [TCA](https://github.com/pointfreeco/swift-composable-architecture) for implementing the chat feature?

Yes it uses an earlier non Combine slightly modified version of the composable architecture.

**Comment by andreyz** [comment]
> > P.S. Are you using [TCA](https://github.com/pointfreeco/swift-composable-architecture) for implementing the chat feature?
> 
> Yes it uses an earlier non Combine slightly modified version of the composable architecture.

Then you would understand what a pain it is to use a singleton bound HSBeacon, that also seems to stop delivering delegate callbacks if I call open HSBeaconSetting with the same beaconId more than once. I had to keep a private global just for working around that.

On a bright side, if you'd like me to share a BeaconState/View wrapper I accidentally made, let me know.

**Comment by ScottPetit** [comment]
Hey @andreyz,

First off thanks for the bug report, I was only able to reproduce this on iOS 14, but logged an [internal ticket](https://helpscout.atlassian.net/browse/IOSSDK-340) to let you know when it's resolved.  

Secondly circling back to TCA.  So an unfortunate side effect of supporting iOS 11 is we're using TCA with UIKit and not SwiftUI.  I was wondering if you could walk me through how you're implementing Beacon, as that comment and one you left on a [previous thread](https://github.com/helpscout/beacon-ios-sdk/issues/73#issuecomment-685762725) struck me as being different from the norm and that seems to be leading to you running into some issues.  Are you calling `dismissBeacon`?  Are you toggling `@State` to handle presenting Beacon/dismissing Beacon?  If so is that how you're leveraging the `HSBeaconDelegate` callbacks?

Thanks.

**Comment by ScottPetit** [comment]
Hey 👋 @andreyz,

Thanks for reporting this, just wanted to write and let you know that this is resolved with the latest version (2.0.2) of the SDK.  If you'd like to see what else is in this release take a look at the [Changelog](https://github.com/helpscout/beacon-ios-sdk/blob/master/CHANGELOG.md).

Also if you have the time to take a look at some of the questions I raised [above](https://github.com/helpscout/beacon-ios-sdk/issues/105#issuecomment-689060622) that extra context could really help us going forward.
